# Coding Exercises (Part 6)

## Pandas: Intermediate

### Exercise 6: NA Values & Summary Statistics

In [1]:
# run the cell!
import pandas as pd
pd.options.display.float_format = '{:.2f}'.format

In [2]:
# run the cell!
listings = pd.read_csv("listings.csv")

In [3]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   ADR TSO  IPO_Year            Sector  \
0      NaN   1999.00     Capital Goods   
1      NaN   2016.00  Basic Industries   
2      NaN       NaN        Technology   
3      NaN   2014.00       Health Care   
4      NaN       NaN    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [4]:
# run the cell
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6745 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   ADR TSO     140 non-null    float64
 6   IPO_Year    3105 non-null   float64
 7   Sector      5309 non-null   object 
 8   Industry    5309 non-null   object 
dtypes: float64(4), object(5)
memory usage: 481.9+ KB


51. Get the __amount__ of __NA-values__ per __column__! __Fill in the gaps__! <br> How many __NA-values__ are in the __ADR TSO__ column?

In [14]:
listings.isna().sum()

Symbol           0
Exchange         0
Name             0
Last_Price     107
Market_Cap     898
ADR TSO       6712
IPO_Year      3747
Sector        1543
Industry      1543
dtype: int64

There are ... 6712 NA-values in the ADR TSO column!

52. Get the __amount__ of __non-null values__ per column (don´t use info() method)! <br>

In [11]:
listings.notnull().sum()

Symbol        6852
Exchange      6852
Name          6852
Last_Price    6745
Market_Cap    5954
ADR TSO        140
IPO_Year      3105
Sector        5309
Industry      5309
dtype: int64

In [12]:
listings.isnull().sum()

Symbol           0
Exchange         0
Name             0
Last_Price     107
Market_Cap     898
ADR TSO       6712
IPO_Year      3747
Sector        1543
Industry      1543
dtype: int64

53. __Drop__ all __rows__ with __at least one NA-Value__ from listings (don´t overwrite listings!). __How many__ rows/listing are __left__?

In [10]:
listings.dropna()
# dropna() method does not change the dataframe by default

Symbol Exchange                                 Name  Last_Price  \
75     ADAP   NASDAQ         Adaptimmune Therapeutics plc        4.56   
305    AMCN   NASDAQ                  AirMedia Group Inc.        0.47   
354    ANCN   NASDAQ           Anchiano Therapeutics Ltd.        6.67   
398    APOP   NASDAQ           Cellect Biotechnology Ltd.        0.99   
449    ARGX   NASDAQ                            argenx SE      127.51   
...     ...      ...                                  ...         ...   
6792     YI   NASDAQ                            111, Inc.        7.10   
6793    YIN   NASDAQ  Yintech Investment Holdings Limited        6.05   
6810     YY   NASDAQ                              YY Inc.       85.55   
6822   ZEAL   NASDAQ                   Zealand Pharma A/S       19.01   
6835   ZLAB   NASDAQ                      Zai Lab Limited       32.43   

        Market_Cap     ADR TSO  IPO_Year       Sector  \
75    366034829.76 80270796.00   2015.00  Health Care   
305     4245804.95  9035550.00   2007.00   Technology   
354    17690000.58  2652174.00   2019.00  Health Care   
398     6432765.64  6468992.00   2016.00  Health Care   
449  2656424245.66 20833066.00   2017.00  Health Care   
...            ...         ...       ...          ...   
6792   56690340.50  7984555.00   2018.00  Health Care   
6793   97220323.75 16069475.00   2016.00      Finance   
6810 4235790354.15 49512453.00   2012.00   Technology   
6822   59540954.86  3132086.00   2017.00  Health Care   
6835  674036535.36 20784352.00   2017.00  Health Care   

                                               Industry  
75    Biotechnology: Biological Products (No Diagnos...  
305                                         Advertising  
354                               Major Pharmaceuticals  
398                          Medical/Dental Instruments  
449   Biotechnology: Biological Products (No Diagnos...  
...                                                 ...  
6792                           Medical/Nursing Services  
6793                 Investment Bankers/Brokers/Service  
6810                                       EDP Services  
6822                              Major Pharmaceuticals  
6835                              Major Pharmaceuticals  

[100 rows x 9 columns]

There are only ... 100 rows left.

54. __Drop__ all __columns__ with __at least one NA-Value__ from listings (don´t overwrite listings!). __How many__ columns are __left__?

In [20]:
listings.dropna(axis=1)

Symbol Exchange                           Name
0         A     NYSE     Agilent Technologies, Inc.
1        AA     NYSE              Alcoa Corporation
2      AABA   NASDAQ                    Altaba Inc.
3       AAC     NYSE             AAC Holdings, Inc.
4       AAL   NASDAQ  American Airlines Group, Inc.
...     ...      ...                            ...
6847   ZUMZ   NASDAQ                    Zumiez Inc.
6848    ZUO     NYSE                    Zuora, Inc.
6849   ZYME     NYSE                 Zymeworks Inc.
6850   ZYNE   NASDAQ  Zynerba Pharmaceuticals, Inc.
6851   ZYXI   NASDAQ                    Zynex, Inc.

[6852 rows x 3 columns]

There are only ... 3 columns left!

55. We finally decide to __drop__ all __columns__ with __less than 3000 non-null values__. __Overwrite__ listings! __Fill in the gaps!__

In [43]:
listings.dropna(axis=1, thresh=3000,inplace=True)

In [26]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6745 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    3105 non-null   float64
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(3), object(5)
memory usage: 428.4+ KB


56. __Filter__ listings for all rows/listings with __NA-values__ in the __Last_Price__ column. Save the new DataFrame in the variable __no_price__. <br>
__Fill in the gaps__!

In [28]:
no_price = listings[listings["Last_Price"].isna()]

What do all listings with __NA-values in the Last_Price__ column have in common (hint: check the __Market_Cap__ column)?

In [29]:
# run the cell and inspect!
no_price.head()

Symbol Exchange                                       Name  Last_Price  \
151   AGM.A     NYSE  Federal Agricultural Mortgage Corporation         NaN   
190  AIG.WS     NYSE         American International Group, Inc.         NaN   
224   AKO.A     NYSE                  Embotelladora Andina S.A.         NaN   
225   AKO.B     NYSE                  Embotelladora Andina S.A.         NaN   
254    ALIT   NASDAQ                                Alight Inc.         NaN   

     Market_Cap  IPO_Year Sector Industry  
151        0.00       NaN    NaN      NaN  
190        0.00       NaN    NaN      NaN  
224        0.00       NaN    NaN      NaN  
225        0.00       NaN    NaN      NaN  
254        0.00       NaN    NaN      NaN

In [30]:
# run the cell and inspect!
no_price.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 107 entries, 151 to 6839
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      107 non-null    object 
 1   Exchange    107 non-null    object 
 2   Name        107 non-null    object 
 3   Last_Price  0 non-null      float64
 4   Market_Cap  107 non-null    float64
 5   IPO_Year    27 non-null     float64
 6   Sector      3 non-null      object 
 7   Industry    3 non-null      object 
dtypes: float64(3), object(5)
memory usage: 7.5+ KB


In [31]:
#run the cell and inspect!
no_price.Market_Cap.value_counts()

0.00    107
Name: Market_Cap, dtype: int64

The Market_Cap is 0.0 for all listings with NA-values in the Last_Price column. Therefore, it might make sense to...

57. ...__Fill all NA-Values__ in the __Last_Price__ column of listings with __0.0__! __Save the change__!

In [44]:
# listings.fillna(0.0,inplace=True)
listings["Last_Price"].fillna(0.0,inplace=True)

In [45]:
# run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    3105 non-null   float64
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(3), object(5)
memory usage: 428.4+ KB


58. __Filter__ listings for all rows/listings that have __non-null values__ in the __IPO_Year__ column! __Save__ the new DataFrame in the variable __known_year__. <br> 
__Fill in the gaps!__

In [46]:
known_year = listings[listings["IPO_Year"].notnull()]
# notna()判断是否非空

__How many__ listings with known IPO_Year are in the __known_year__ DataFrame?

In [47]:
# run the cell!
known_year.head()

Symbol Exchange                           Name  Last_Price      Market_Cap  \
0       A     NYSE     Agilent Technologies, Inc.       81.68  25934696179.92   
1      AA     NYSE              Alcoa Corporation       29.15   5407809589.60   
3     AAC     NYSE             AAC Holdings, Inc.        2.16     53141086.80   
8    AAOI   NASDAQ  Applied Optoelectronics, Inc.       12.30    244556139.90   
11   AAPL   NASDAQ                     Apple Inc.      200.62 945979473600.00   

    IPO_Year            Sector  \
0    1999.00     Capital Goods   
1    2016.00  Basic Industries   
3    2014.00       Health Care   
8    2013.00        Technology   
11   1980.00        Technology   

                                            Industry  
0   Biotechnology: Laboratory Analytical Instruments  
1                                           Aluminum  
3                               Medical Specialities  
8                                     Semiconductors  
11                            Computer Manufacturing

In [48]:
# run the cell!
known_year.shape

(3105, 8)

There are ... 3105 listings with known IPO_Year!

Finally, we figured out that __IPO_Year__ information is __not available__ for stocks with an IPO __prior to 1970__. Therefore...

59. ... we decide to __fill all NA-Values__ in the __IPO_Year__ column with __1969__. __Save__ the change!

In [50]:
listings.IPO_Year.fillna(1969,inplace=True)

Make a __final inspection__ of our listings DataFrame. Which __columns__ still contain __NA-values__?

In [51]:
# run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year            Sector  \
0   1999.00     Capital Goods   
1   2016.00  Basic Industries   
2   1969.00        Technology   
3   2014.00       Health Care   
4   1969.00    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [52]:
#run the cell!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    6852 non-null   float64
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(3), object(5)
memory usage: 428.4+ KB


The columns ... Sector & Industry still contain NA-values!

60. __Transform the datatype__ in the IPO_Year column to integer (beyond the scope of this course, run the cell!)

In [53]:
#run the cell!
listings.IPO_Year = listings.IPO_Year.astype("int")

In [54]:
#run the cell and inspect!
listings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    6852 non-null   int64  
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 428.4+ KB


In [55]:
#run the cell!
listings.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year            Sector  \
0      1999     Capital Goods   
1      2016  Basic Industries   
2      1969        Technology   
3      2014       Health Care   
4      1969    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

61. __Save and export__ the cleaned DataFrame to the new csv-file __listings_clean.csv__! __Fill in the gaps!__

In [56]:
listings.to_csv("listings_clean.csv", index = False)

62. __Re-Import__ listings_clean.csv and save the DataFrame in the variable __listings_clean__!

In [57]:
listings_clean=pd.read_csv("listings_clean.csv")

In [58]:
# run the cell!
listings_clean.head()

Symbol Exchange                           Name  Last_Price     Market_Cap  \
0      A     NYSE     Agilent Technologies, Inc.       81.68 25934696179.92   
1     AA     NYSE              Alcoa Corporation       29.15  5407809589.60   
2   AABA   NASDAQ                    Altaba Inc.       75.39 42781131315.30   
3    AAC     NYSE             AAC Holdings, Inc.        2.16    53141086.80   
4    AAL   NASDAQ  American Airlines Group, Inc.       34.02 15276869742.96   

   IPO_Year            Sector  \
0      1999     Capital Goods   
1      2016  Basic Industries   
2      1969        Technology   
3      2014       Health Care   
4      1969    Transportation   

                                           Industry  
0  Biotechnology: Laboratory Analytical Instruments  
1                                          Aluminum  
2                                      EDP Services  
3                              Medical Specialities  
4                     Air Freight/Delivery Services

In [59]:
# run the cell!
listings_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6852 entries, 0 to 6851
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Symbol      6852 non-null   object 
 1   Exchange    6852 non-null   object 
 2   Name        6852 non-null   object 
 3   Last_Price  6852 non-null   float64
 4   Market_Cap  5954 non-null   float64
 5   IPO_Year    6852 non-null   int64  
 6   Sector      5309 non-null   object 
 7   Industry    5309 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 428.4+ KB


63. Call an appropriate method on the __listings_clean__ DataFrame that provides plenty of __summary statistics__!

In [60]:
listings_clean.describe()

Last_Price      Market_Cap  IPO_Year
count     6852.00         5954.00   6852.00
mean        36.79   7174440219.09   1987.14
std         98.58  35254552925.59     20.89
min          0.00            0.00   1969.00
25%          7.70    120376889.75   1969.00
50%         19.98    558517593.27   1969.00
75%         36.85   2888746939.46   2012.00
max       4299.99 945979473600.00   2019.00

63a. __Calculate__ the __mean__ and the __standard deviation__ for the numerical columns (use the appropriate method!).

In [65]:
listings_clean.aggregate(["mean","std"])
# aggregate自动将统计函数用到类型为数值的栏目

Last_Price     Market_Cap  IPO_Year
mean       36.79  7174440219.09   1987.14
std        98.58 35254552925.59     20.89

64. Get some summary statistics on all __non-numerical columns__ by passing __"O"__ to the __include parameter__ of the __describe()__ method! __Fill in the gaps!__ <br>
What is the __most frequent__ value in the __Industry__ column? 

In [72]:
listings_clean.describe(include = "O")
# listings.describe?
# listings_clean.describe()
# describe默认对数值栏目进行统计，正如aggregate一样。如果想要对非数值栏目统计，则要include="O"

Symbol Exchange            Name   Sector               Industry
count    6852     6852            6852     5309                   5309
unique   6840        3            5766       12                    135
top      CDOR   NASDAQ  Public Storage  Finance  Major Pharmaceuticals
freq        2     3435              14     1059                    470

The most frequent Industry is ... Major Pharmaceuticals with 470 appearances.

65. Get the __maximum__ values for all columns! Which is the (alphabetically) __last Symbol__?

In [75]:
listings_clean.max()

/var/folders/b8/lmz1tyx9785389sr236542740000gn/T/ipykernel_1810/3025616438.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  listings_clean.max()


Symbol                  ZYXI
Exchange                NYSE
Name             voxeljet AG
Last_Price           4299.99
Market_Cap   945979473600.00
IPO_Year                2019
dtype: object

The (alphabetically) last Symbol is ... ZYXI!

66. __Calculate__ the __total Market_Cap__ over all listings.

In [76]:
listings.Market_Cap.sum()

42716617064473.484

67. Create a __correlation matrix__ for the __listings_clean__ DataFrame! Are Market_Cap and Last_Price __positively correlated__?

In [77]:
listings_clean.corr()

Last_Price  Market_Cap  IPO_Year
Last_Price        1.00        0.31     -0.06
Market_Cap        0.31        1.00     -0.07
IPO_Year         -0.06       -0.07      1.00

Market_Cap and Last_Price are ... positively correlated.The higher the Price, the higher the Market Capitalization.

------------------------------

# Recap

51. count __na__ items

`df.isna().sum()`

52. count __non__ na items

`df.notna().sum()`

55. dropna and only show columns with __more__ than N items

`df.dropna(thresh=N)`

57. fill na items with specific values

`df.fillna('')` or `df.fillna(method=)`

63. conduct specific statistical insights of the dataframe

`df.aggregate(['mean','std'])`

64. get statistical summary of the dataframe(__including__ non numeric columns)

`df.describe(including='O')`